fix: delete parent meta-relation if needed and support undo

yinanazhou opened the pull request:
- Added `find_parent_meta_relation`
- Rename `metarel` to `is_meta_relation`
- Use `let` and `const` instead of `var` for best practice

refs: #63

yrammos commented on src/js/delete.js:
```diff
@@ -16,37 +14,51 @@ function delete_relation(elem) {
-              arc.getAttribute('from') == '#' + elem.id
-    })
-  var removed = arcs.concat(svg_hes)
+  // Find all arcs related to this element
+  const arcs = Array.from(mei.getElementsByTagName('arc')).filter(arc => {
+    return arc.getAttribute('to') == '#' + elem.id || arc.getAttribute('from') == '#' + elem.id
+  })
```
Are these three lines actually different (apart from formatting)? If not, please revert to original formatting for history purposes.

yrammos commented on src/js/delete.js:
```diff
@@ -55,11 +67,18 @@ function delete_relation(elem) {
 
 export function delete_relations(redoing = false) {
-  console.debug('Using globals: selected for element selection, undo_actions for storing the action')
+  console.debug(
+    'Using globals: selected for element selection, undo_actions for storing the action'
+  )
```
Ibid.  Pls discard format-only changes.

yrammos commented on src/js/delete.js:
```diff
@@ -55,11 +67,18 @@ function delete_relation(elem) {
   var sel = selected.concat(extraselected)
-  if (sel.length == 0 || !(get_class_from_classlist(sel[0]) == 'relation' ||
-	                  get_class_from_classlist(sel[0]) == 'metarelation')) {
+  if (
+    sel.length == 0 ||
+    !(
+      get_class_from_classlist(sel[0]) == 'relation' ||
+      get_class_from_classlist(sel[0]) == 'metarelation'
+    )
+  ) {
```
Ibid.

yrammos commented on src/js/delete.js:
```diff
@@ -68,6 +87,62 @@ export function delete_relations(redoing = false) {
+ */
+function find_parent_meta_relations(elem, mei, draw_contexts, svg_hes) {
+  let meta_relations = []
+  let meta_relation_arcs = []
+
+  // Find arcs that connect meta-relations to this relation
+  const meta_arcs = Array.from(mei.getElementsByTagName('arc')).filter(
+    arc => arc.getAttribute('to') == '#' + elem.id || arc.getAttribute('from') == '#' + elem.id
+  )
+
+  // For each arc, find the meta-relation nodes
+  meta_arcs.forEach(arc => {
+    // Find the meta-relation (whether it's at the "to" or "from" end)
+    const meta_id =
+      arc.getAttribute('to') === '#' + elem.id
+        ? arc.getAttribute('from').substring(1)
+        : arc.getAttribute('to').substring(1)
+
+    const meta_node = get_by_id(mei, meta_id)
+    if (meta_node && meta_node.getAttribute('type') === 'metarelation') {
+      meta_relations.push(meta_node)
+
+      // Find SVG elements for this meta-relation
+      for (const draw_context of draw_contexts) {
+        let svg_meta = get_by_id(document, draw_context.id_prefix + meta_id)
+        if (svg_meta) {
+          svg_hes.push(svg_meta)
+        }
+      }
+
+      // Find all arcs related to this meta-relation
+      const related_arcs = Array.from(mei.getElementsByTagName('arc')).filter(
+        a => a.getAttribute('to') == '#' + meta_id || a.getAttribute('from') == '#' + meta_id
+      )
+
+      meta_relation_arcs = meta_relation_arcs.concat(related_arcs)
+    }
+  })
+
+  return { meta_relations, meta_relation_arcs }
```
Some meta-relations consist of relations. But others consist of meta-relations! There can be, in principle, an arbitrary number of "meta" levels (as demoed on Zoom). So the `find_pattern_meta_relations` method function should somehow work recursively. He can continue the discussion on Mattermost if you want.

yrammos reviewed (changes requested):
Great work (I also appreciate the elegance and comment quality).

But the PR does not address the general case of meta-meta-(…)-meta-relations. Do you want to address it in this PR? Doing so seems more logical to me, rather than treating the general case as a separate issue.

yinanazhou committed:
fix: delete parent meta-relation if needed and support undo

- Added `find_parent_meta_relation`
- Rename `metarel` to `is_meta_relation`
- Use `let` and `const` instead of `var` for best practice

refs: #63

yinanazhou commented:
> Great work (I also appreciate the elegance and comment quality).
> 
> But the PR does not address the general case of meta-meta-(…)-meta-relations. Do you want to address it in this PR? Doing so seems more logical to me, rather than treating the general case as a separate issue.

I prefer to address this in another PR as it is not described in the related issue. I will raise another issue about making it recursive.

yrammos commented:
Okay, in that case, let’s merge this in (after reverting the three spurious hunks, see my comments above).

yrammos commented:
@yinanazhou I reviewed your code, and it looks good to merge (except for the three small points I raise above).

yrammos reviewed (changes requested):
Please revert the formatter-induced diffs I point out in the discussion. Otherwise this looks good.
